odooTimereg.py

```diff
@@ -26,60 +26,50 @@ def login(self, odoouri, user, password, db):
             password=self.password)
 
     def projects(self, client):
-        project_model = client.AccountAnalyticAccount
-        ids = project_model.search(['use_timesheets=True', 'invoice_on_timesheets=True', 'state=("open","pending")'])
-        projects = project_model.read(ids, ['name'])
+        project_model = client.ProjectProject
+        project_ids = project_model.search([
+            ('allow_timesheets', '=', True),
+            ('billable_type', '!=', 'no'),
+            ])
+        projects = project_model.read(project_ids, ['name'])
         return projects
 
     def hours(self, client, projectids, from_date=None, to_date=None):
-        # Read all billable types and differentiate between billable and non-billable
-        # using the discount factor (if < 100.0, we consider it billable, since 100%
-        # is non billable).
-        billable = set()
-        for fact in client.Hr_timesheet_invoiceFactor.read([]):
-            if fact["factor"] < 100:
-                billable.add(fact["id"])
-
-        timesheet_model = client.HrAnalyticTimesheet
+        timesheet_model = client.AccountAnalyticLine
         ids = timesheet_model.search([
-            'account_name=%s' % ",".join(['"'+p+'"' for p in projectids]),
-            'date >= %s' % from_date,
-            'date < %s' % to_date,
+            ('project_id', 'in', projectids),
+            ('date', '>=', from_date.strftime('%Y-%m-%d')),
+            ('date', '<', to_date.strftime('%Y-%m-%d')),
+            ('timesheet_invoice_type', '!=', 'non_billable_project'),
         ])
         if not ids:
             return []
 
-        data = timesheet_model.read(ids, ['account_name', 'user_id', 'date', 'line_id', 'to_invoice', 'unit_amount'])
-
+        data = timesheet_model.read(ids, ['project_id', 'user_id', 'date',
+                                          'move_id', 'timesheet_invoice_type',
+                                          'unit_amount'])
         hours = []
         for item in data:
-            if not item['to_invoice']:
-                continue
             h = {}
-            h["project"] = item["account_name"]
+            h["project"] = item["project_id"]
             h["date"] = item["date"]
             h["user"] = item["user_id"][1]
-            h["remark"] = item["line_id"][1]
+            h["remark"] = item["move_id"][1] if item['move_id'] else ''
```

Comment:
@rasky Ho solo un'incertezza su questo campo. Avresti modo di controllare sulla vecchia versione a cosa fa riferimento line_id di hr.analytic.timesheet, per favore?

Reply:
Questo è un record di quella tabella:

```
{'account_id': [833,
                'Progetti /  Unitec 2016 07 HW - Industrializzazione 4 M6'],
 'account_name': ' Unitec 2016 07 HW - Industrializzazione 4 M6',
 'amount': -15.5,
 'amount_currency': 0.0,
 'code': False,
 'company_id': [1, 'Develer SRL'],
 'currency_id': False,
 'date': '2019-12-04',
 'date_aal': '2019-12-04',
 'general_account_id': [1069, '72 05 010 Salari E Stipendi'],
 'id': 430501,
 'invoice_id': False,
 'journal_id': [6, 'Timesheet'],
 'line_id': [783553, 'Organizzazione collaudo'],
 'move_id': False,
 'name': 'Organizzazione collaudo',
 'partner_id': [5687, 'Unitec Spa'],
 'product_id': [3425, 'Sviluppo software - Nicola Sansonne'],
 'product_uom_id': [12, 'Ore'],
 'ref': False,
 'sheet_id': [3000, 'Month 12'],
 'to_invoice': [11, 'Non Billable'],
 'unit_amount': 0.5,
 'user_id': [104, 'Sansonne  Nicola']}```

Io non ho molte altre informazioni se non che quel campo contiene la descrizione della registrazione ore effettuata. Cioè quando l'utente registra le ore lavorate, per ogni registrazione c'è una descrizione che può inserire, e in quel campo io la trovo. Più di questo non saprei...

Reply:
Ah noto ora che la stessa descrizione è presente banalmente in `name`. Non so se all'epoca non l'avevo vista o non c'era. Per me è indifferente ovviamente, basta estrarre quel campo dovunque esso sia.

Reply:
@OpenCode ti torna tutto? Posso procedere al merge o devi cambiare qualcosa vista la mia analisi?

Reply:
@rasky Sposto il valore di name del record in remark del dizionario. Faccio la modifica a breve.

Reply:
@rasky Fatto. Confermo che il campo era name poiché contiene la descrizione dell'attività svolta dall'utente sul timesheet.